core/src/main/java/org/elasticsearch/search/DefaultSearchContext.java

```diff
@@ -291,11 +290,21 @@ public Query searchFilter(String[] types) {
     static Query createSearchFilter(String[] types, Query aliasFilter, boolean hasNestedFields) {
         Query typesFilter = null;
         if (types != null && types.length >= 1) {
-            BytesRef[] typesBytes = new BytesRef[types.length];
-            for (int i = 0; i < typesBytes.length; i++) {
-                typesBytes[i] = new BytesRef(types[i]);
+            final int threshold = Math.min(16, BooleanQuery.getMaxClauseCount());
+            if (types.length < threshold) {
+                BooleanQuery.Builder builder = new BooleanQuery.Builder();
+                for (String type : types) {
+                    BytesRef typeBytes = new BytesRef(type);
+                    builder.add(new TypeFieldMapper.TypeQuery(typeBytes), Occur.SHOULD);
+                }
+                typesFilter = builder.build();
+            } else {
+                BytesRef[] typesBytes = new BytesRef[types.length];
+                for (int i = 0; i < typesBytes.length; i++) {
+                    typesBytes[i] = new BytesRef(types[i]);
+                }
+                typesFilter = new TermsQuery(TypeFieldMapper.NAME, typesBytes);
```

Comment:
I suspect it's not easy but it would be nice to let the mappings create the query rather than doing it manually.